### PR TITLE
Initial implementation of reduced splitting area

### PR DESCRIPTION
**Pull request by enzbus**
Hello, this is an initial implementation of  #150. It passes `cargo build`, `cargo test` in rust, `maturin build` and `pytest` in python. I am not able to compile R, I copied to the best of my ability the code from what was already there. I also tested in python that it does what is supposed to do. I think as a first step it's enough to just limit the splitting area in the way I'm doing here: there are two new control hyperparameters: `nosplit_before_index` and `nosplit_after_index`, both optional, and if they are `None` it behaves as before adding this feature. I haven't added tests yet, please direct me on how to do that. Thanks!

### CODE REVIEW

**Comment by mlondschien** [comment]
Thanks!

I am unsure of what this is supposed to do: I understood #150 as that you'd like to supply a single segment to `changeforest` that is "off-limits". What you implement here however disallows splitting before / after a certain index. I.e., you disallow splitting in the segments `[0, nosplit_before_index)` and (nosplit_after_index, n]`. Is this correct?

A practical consideration: Why not allow supplying a list `forbidden_segments = [(start1, stop1), (start2, stop2), ..., ]`? This would include both use cases.

The question would remain on how to choose the initial guesses in `TwoStepSearch.find_best_split`. A pragmatic approach would be to choose the (unique) entries in `split_candidates` that are closest to the 1/4, 1/2, and 3/4 quantiles of `[start, stop)`. What do you think?

**Comment by enzbus** [comment]
Hello @mlondschien , right, this is not exactly what I suggested in #150. As I was implementing this first cut I realized it made more sense to just limit initial and final indexes, at least for the usecase I'm targeting for this. If you want instead `forbidden_segments`, there would be a few more technical issues. Yes, for `TwoStepSearch` some heuristics like that would work, but I'm more concerned about the glue code required for the R and python interface, correctly handling syntax errors, ... Think about it, this that I coded now is good for my usecase, we can polish this with tests, ..., or do it the other way.

**Comment by mlondschien** [comment]
>  but I'm more concerned about the glue code required for the R and python interface, correctly handling syntax errors

Hmm, which kind of syntax errors do you mean? For the interface, I would expect something like

```
        if let Ok(pyvalue) = obj.getattr(py, "forbidden_segments") {
            if let Ok(value) = pyvalue.extract::<Option<Vec<(usize, usize)>>(py) {
                control = control.with_forbidden_segments(value);
            }
        };
```

to work. The R glue might be a bit more complex, but we can handle this in a second step.

**Comment by enzbus** [comment]
Ok sounds good, I'll rework it. I don't know rust, it looks like it has c++ templates so I should be good with that. And yes I would focus on the python interface first.

**Comment by mlondschien** [comment]
It has been some time since I last touched Rust code. Maybe something like this:


```
/// Vector with indices of allowed split points.
fn split_candidates(&self, start: usize, stop: usize) -> Result<Vec<usize>, &str> {
    let minimal_segment_length =
        (self.control().minimal_relative_segment_length * (self.n() as f64)).ceil() as usize;

    if (actual_stop < actual_start) || (2 * minimal_segment_length >= (actual_stop - actual_start)) {
        Err("Segment too small.")
    } else {
        let mut split_candidates = ((start + minimal_segment_length)..(stop - minimal_segment_length)).collect();

        if let Some(forbidden_segments) = self.control().forbidden_segments {
            split_candidates = split_candidates.filter(|x| forbidden_segments.map(|a, b|, x < a || x >= b).all())
        }

        if split_candidates.len() == 0 {
            Err("No split_candidates left after filtering out forbidden_segments.")
        } else {
            Ok(split_candidates)
        }
    }
}
```
would do the job? It's not the most efficient implementation, but should work.

**Comment by enzbus** [comment]
Hello @mlondschien please review, this is like what you suggested. Let me know where to add tests, I would prefer to do it in Python since I don't know rust. I have tested with the interpreter already and it seems to work. Thanks!

File: src/optimizer/optimizer.rs [hunk 1]
```diff
@@ -23,7 +23,20 @@ pub trait Optimizer {
         if 2 * minimal_segment_length >= (stop - start) {
             Err("Segment too small.")
         } else {
-            Ok(((start + minimal_segment_length)..(stop - minimal_segment_length)).collect())
+            // implementation suggested by @mlondschien in https://github.com/mlondschien/changeforest/pull/151
```
Comment (mlondschien) [inline src/optimizer/optimizer.rs:26]:
```suggestion
```

File: src/optimizer/two_step_search.rs [hunk 2]
```diff
@@ -51,13 +51,33 @@ where
...
+            let cand = split_candidates.clone().into_iter().nth(split_candidates.len() / 2).unwrap();
+            if cand > guesses[guesses.len()-1] {guesses.push(cand)};
+
+            // same
+            let cand = split_candidates.clone().into_iter().nth(3 * split_candidates.len() / 4).unwrap();
+            if cand > guesses[guesses.len()-1] {guesses.push(cand)};
+
+            } else {
+
+            // this is same logic that was there before
```
Comment (mlondschien) [inline src/optimizer/two_step_search.rs:74]:
```suggestion
```

**Review by mlondschien — COMMENTED** [review]
Looks good, thanks! I would like to add some rust tests as well. Can I push to your branch?

**Comment by enzbus** [comment]
> Looks good, thanks! I would like to add some rust tests as well. Can I push to your branch?

Ok, sure, I added you w/ write access on my forked repo (I think if you push to that it adds to this merge request?)

**Comment by enzbus** [comment]
That's good with me, thanks!

**Pull request merged**